presto-main/src/main/java/com/facebook/presto/sql/relational/optimizer/ExpressionOptimizer.java

```diff
@@ -89,12 +97,42 @@ public RowExpression visitConstant(ConstantExpression literal, Void context)
         @Override
         public RowExpression visitCall(CallExpression call, Void context)
         {
-            ScalarFunctionImplementation function;
+            ScalarFunctionImplementation function = null;
             Signature signature = call.getSignature();
 
             if (signature.getName().equals(CAST)) {
-                Signature functionSignature = registry.getCoercion(call.getArguments().get(0).getType(), call.getType());
-                function = registry.getScalarFunctionImplementation(functionSignature);
+                Signature functionSignature = null;
+                if (call.getArguments().get(0) instanceof CallExpression) {
+                    // Optimization for CAST(JSON_PARSE(...) AS ARRAY/MAP)
+                    CallExpression innerCall = (CallExpression) call.getArguments().get(0);
+                    if (innerCall.getSignature().getName().equals("json_parse")) {
+                        checkState(innerCall.getType().equals(JSON));
```

Comment:
These `checkState`s are more like assertions, so I guess `Verify.verify()` is a better fit here.